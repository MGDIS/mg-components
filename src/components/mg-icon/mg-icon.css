:host {
  display: block;
}

.icon {
  display: inline-block;
  color: var(--color-icon-primary);
  height: 1em;
  width: 1em;
  line-height: 1;
}
