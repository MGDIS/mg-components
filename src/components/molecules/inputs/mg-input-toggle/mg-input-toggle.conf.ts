/**
 * type ToggleValue
 * use to match toggle input attributes
 */
export type ToggleValue = {
  title: string;
  value: unknown;
};
