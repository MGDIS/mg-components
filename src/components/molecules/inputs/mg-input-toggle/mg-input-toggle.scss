@use '../styles/mg-input.scss';
@use '../../../../styles/a11y.scss';

/**
* Toogle
*/

// variables
$color-light-blue: hsl(var(--color-info-h) calc(var(--color-info-s) - 64%) calc(var(--color-info-l) + 55%));
$color-light-dark: hsl(var(--color-dark-h) calc(var(--color-dark-s) - 30%) calc(var(--color-dark-l) + 75%));
$button-toggle-padding: 0.2rem;
$toggle-item-container-padding-vertical: 0.5rem;

.mg-input.mg-input.mg-input--toggle .mg-input__input-container {
  flex: auto;
  min-width: 0;
}

.mg-input__button-toggle {
  // Display
  display: flex;
  padding: $button-toggle-padding;
  min-width: 0;
  // Decoration
  background-color: hsl(var(--color-light));
  border-width: var(--mg-inputs-border-width);
  border-style: solid;
  border-color: var(--mg-inputs-color);
  border-radius: calc(
    ((var(--font-size) * var(--line-height)) + (2 * ($button-toggle-padding + $toggle-item-container-padding-vertical + var(--mg-inputs-border-width)))) /
      var(--mg-input-toggle-border-radius-ratio)
  );
  // Font
  font-family: var(--font-family);
  font-size: inherit;
  line-height: inherit;

  cursor: pointer;
  &:disabled {
    cursor: default;
  }
}

.mg-input__toggle-item-container {
  padding: $toggle-item-container-padding-vertical 0.9rem;
  border: 0;
  box-sizing: border-box;
  border-radius: calc(((var(--font-size) * var(--line-height)) + (2 * $toggle-item-container-padding-vertical)) / var(--mg-input-toggle-border-radius-ratio));

  background-color: unset;
  font-weight: unset;
  color: hsl(var(--color-dark));

  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;

  &:first-of-type {
    font-weight: 600;
    color: hsl(var(--color-info));
    background-color: $color-light-blue;
  }
}

// toggle default active variation
.mg-input.mg-input--toggle-is-active {
  .mg-input__toggle-item-container {
    &:first-of-type {
      background-color: unset;
      font-weight: unset;
      color: hsl(var(--color-dark));
    }
    &:last-of-type {
      font-weight: 600;
      color: hsl(var(--color-info));
      background-color: $color-light-blue;
    }
  }
}

// icon
.mg-input.mg-input--toggle-icon .mg-input__toggle-item-container {
  display: flex;
  justify-content: center;
  align-items: center;
  width: calc(((var(--font-size) * var(--line-height)) + (2 * $toggle-item-container-padding-vertical)));
  height: calc(((var(--font-size) * var(--line-height)) + (2 * $toggle-item-container-padding-vertical)));
  padding: 0;
  visibility: hidden;

  &:first-of-type {
    visibility: visible;
  }

  ::slotted(mg-icon) {
    display: flex;
  }
}

.mg-input.mg-input--toggle-icon.mg-input--toggle-is-active .mg-input__toggle-item-container {
  &:first-of-type {
    visibility: hidden;
  }

  &:last-of-type {
    visibility: visible;
  }
}

// on/off
.mg-input.mg-input--toggle-on-off {
  .mg-input__toggle-item-container:first-of-type {
    font-weight: 600;
    color: hsl(var(--color-dark));
    background-color: $color-light-dark;
  }

  &.mg-input--toggle-is-active .mg-input__toggle-item-container:first-of-type {
    background-color: unset;
    font-weight: unset;
    color: hsl(var(--color-dark));
  }
}

// readonly
.mg-input.mg-input--toggle-readonly {
  .mg-input__button-toggle {
    border-color: transparent;
    background-color: transparent;
    cursor: initial;
  }

  .mg-input__toggle-item-container:last-of-type {
    display: none;
  }

  &.mg-input--toggle-is-active .mg-input__toggle-item-container {
    &:first-of-type {
      display: none;
    }

    &:last-of-type {
      display: flex;
    }
  }
}
