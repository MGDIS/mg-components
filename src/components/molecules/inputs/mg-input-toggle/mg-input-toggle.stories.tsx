import { h } from "@stencil/core";

export default {
  component: 'mg-input-toggle',
  title: 'Molecules/Inputs/mg-input-toggle',
};

const args = {
  // Global
  value: null,
  items: [{title: 'non', value: false}, {title: 'oui', value: true}],
  identifier: 'identifier',
  name: 'input-name',
  // Label
  label: 'Option',
  labelOnTop: false,
  labelHide: false,
  // toggle
  displayBothValues: false,
  // Input
  required: false,
  disabled: false,
  readonly: false,
  // Tooltip
  tooltip: 'This is a tooltip',
  // Help Text
  helpText: 'Help text with html <strong>bold</strong>, <em>italic</em>.'
}

/**
 * 1. camelCase arguments must be written in the template, for exemple labelOnTop must be placed in the template as label-on-top={args.labelOnTop}
 * 2. boolean arguments with a default true value must be added like display-character-left={args.displayCharacterLeft ? 'true' : 'false'}
 */
const Template = args => {
  // Extract slot so it won't be render as an attribute
  const labelOnTop = args.labelOnTop;
  delete args.labelOnTop;
  const labelHide = args.labelHide;
  delete args.labelHide;
  const displayBothValues = args.displayBothValues;
  delete args.displayBothValues;
  const helpText = args.helpText;
  delete args.helpText;
  // return element
  return   <mg-input-toggle
    {...args}
    label-on-top={labelOnTop}
    label-hide={labelHide}
    help-text={helpText}
    display-both-values={displayBothValues}
  >
    <span slot="item-1">Non</span>
    <span slot="item-2">Oui</span>
  </mg-input-toggle>
};

export const MgInputtoggle = Template.bind({});
MgInputtoggle.args = {...args };

const TemplateIcon = args => {
  // Extract slot so it won't be render as an attribute
  const labelOnTop = args.labelOnTop;
  delete args.labelOnTop;
  const helpText = args.helpText;
  delete args.helpText;
  const labelHide = args.labelHide;
  delete args.labelHide;
  const displayBothValues = args.displayBothValues;
  delete args.displayBothValues;
  // return element
  return   <mg-input-toggle
    {...args}
    label-on-top={labelOnTop}
    label-hide={labelHide}
    help-text={helpText}
    display-both-values={displayBothValues}
  >
    <span slot="item-1" class="mg-input-toggle__item-icon"><mg-icon icon="cross"></mg-icon></span>
    <span slot="item-2" class="mg-input-toggle__item-icon"><mg-icon icon="check-circle"></mg-icon></span>
  </mg-input-toggle>
};

export const MgInputtoggleWithIcon = TemplateIcon.bind({});
MgInputtoggleWithIcon.args = {...args};
