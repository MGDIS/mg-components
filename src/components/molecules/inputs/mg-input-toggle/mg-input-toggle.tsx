import { Component, Event, h, Prop, EventEmitter, State, Watch } from '@stencil/core';
import { MgInput } from '../MgInput';
import { createID, ClassList, allItemsAreString } from '../../../../utils/components.utils';
import { ToggleValue } from '../../../../types/components.types'
import { messages } from '../../../../locales';

/**
* type Option validation function
* @param option
* @returns {boolean}
*/
const isOption = (option: ToggleValue): boolean => typeof option === 'object' && typeof option.title === 'string' && option.value !== undefined;


@Component({
  tag: 'mg-input-toggle',
  styleUrl: 'mg-input-toggle.scss',
  scoped: true,
})
export class MgInputToggle {

  /************
   * Internal *
   ************/

  private classError = 'is-not-valid';
  private classToggleActive = 'mg-input--toggle-active';

  /**************
   * Decorators *
   **************/

  /**
  * Component value
  */
  @Prop({ mutable:true, reflect: true }) value?: any;

  /**
  * Items are the possible options to select
  * Required
  */
  @Prop() items!: string[] | ToggleValue[];
  @Watch('items')
  validateItems(newValue){
    if (typeof newValue === 'object' && this.items.length !== 2) {
      throw new Error('<mg-input-toggle> prop "items" require 2 items.')
    }
    // String array
    else if(allItemsAreString(newValue)) {
      this.options = newValue.map(item => ({ title:item, value:item }));
    }
    // Object array
    else if(newValue && (newValue as Array<ToggleValue>).every(item => isOption(item))) {
      this.options = newValue;
    }
    else {
      throw new Error('<mg-input-toggle> prop "items" is required and all items must be the same type: ToggleValue.')
    }
  }

  /**
  * Identifier is used for the element ID (id is a reserved prop in Stencil.js)
  * If not set, it will be created.
  */
  @Prop() identifier?: string = createID('mg-input-toggle');

  /**
  * Input name
  * If not set the value equals the identifier
  */
  @Prop() name?: string = this.identifier;

  /**
  * Input label
  * Required
  */
  @Prop() label!: string;

  /**
   * Define if label is displayed on top
   */
  @Prop() labelOnTop: boolean;

  /**
   * Define if label is visible
   */
  @Prop() labelHide: boolean = false;

  /**
  * Define if values are both displayed or only active one
  */
  @Prop() displayBothValues?: boolean = false;
  @Watch('displayBothValues')
  validateDisplayBothValues(newValue: boolean) {
    if(newValue) this.classList.add(`mg-input-toggle--display-both-values`);
  }

  /**
  * Define if input is required
  */
  @Prop() required: boolean = false;

  /**
  * Define if input is readonly
  */
  @Prop() readonly: boolean = false;

  /**
  * Define if input is disabled
  */
  @Prop() disabled: boolean = false;

  /**
  * Add a tooltip message next to the input
  */
  @Prop() tooltip: string;

  /**
  * Template to use for characters left sentence
  */
  @Prop() helpText: string;

  /**
  * Force valid component
  */
  @Prop({ mutable: true, reflect: true }) valid: boolean;

  /**
  * Force invalid component
  */
  @Prop({ mutable: true, reflect: true }) invalid: boolean;

  /**
  * Component classes
  */
  @State() classList: ClassList = new ClassList(['mg-input--toggle']);

  /**
  * Formated items for display
  */
  @State() options: ToggleValue[];

  /**
  * Error message to display
  */
  @State() errorMessage: string;

  /**
  * Emmited event when value change
  */
  @Event() valueChange: EventEmitter<any>

  private toggleValue = () => {
    this.value = this.value === undefined
      ? this.options[1].value
      : this.options.find(o => o.value !== this.value).value;

    this.valueChange.emit(this.value);
    this.setToggleStyle()
  }

  private setToggleStyle() {
    if (this.value !== undefined && this.value !== this.options[0].value) {
      this.classList.add(this.classToggleActive);
    } else {
      this.classList.delete(this.classToggleActive);
    }
    this.classList = new ClassList(this.classList.classes);
  }

  /**
  * Handle input event
  * @param event
  */
  private handleToggleClick = () => {
    this.toggleValue();
  }

  private handleToggleKeyboard = (event: KeyboardEvent) => {
    if (['ArrowRight', 'ArrowLeft', 'ArrowUp', 'ArrowDown'].includes(event.code)) {
      this.toggleValue();
    }
  }

  /**
  * Handle blur event
  * @param event
  */
  private handleBlur = () => {
    // Manage focus
    this.classList = new ClassList(this.classList.classes);

    // Check validity
    this.checkValidity();
  }

  /**
  * Check if input is valid
  * @param element
  */
  private checkValidity() {
    const valueMissing :boolean = this.value === undefined ? true : false;

    // Set error message
    this.errorMessage = undefined;
    if(valueMissing) {
      this.errorMessage = messages.errors.required;
    }

    // Set validity
    this.valid = valueMissing === false;
    this.invalid = !this.valid;

    // Update class
    if(this.valid) {
      this.classList.delete(this.classError);
    }
    else {
      this.classList.add(this.classError);
    }
  }

  /*************
  * Lifecycle *
  *************/

  /**
  * Check if component props are well configured on init
  */
   componentWillLoad() {
    // Check items format
    this.validateItems(this.items);

    this.setToggleStyle();
    this.validateDisplayBothValues(this.displayBothValues)
  }

  render() {
    return (<MgInput
      identifier={this.identifier}
      classList={this.classList}
      label={this.label}
      labelOnTop={this.labelOnTop}
      labelHide={this.labelHide}
      required={this.required}
      readonly={this.readonly}
      value={this.value?.toString()}
      readonlyValue={undefined}
      tooltip={this.tooltip}
      displayCharacterLeft={undefined}
      characterLeftTemplate={undefined}
      maxlength={undefined}
      helpText={this.helpText}
      errorMessage={this.errorMessage}
      isFieldset={false}
    >
        <button
          type="button"
          role="switch"
          aria-checked={this.value}
          id={this.identifier}
          class={`mg-input-toggle__button ${this.disabled ? 'mg-input-toggle__button--disabled' : ''}`}
          disabled={this.disabled}
          onBlur={this.handleBlur}
          onClick={this.handleToggleClick}
          onKeyDown={this.handleToggleKeyboard}
        >
          <span aria-hidden="true" class="mg-input-toggle__item-container">
            <slot name="item-1"></slot>
          </span>
          <span aria-hidden="true" class="mg-input-toggle__item-container">
            <slot name="item-2"></slot>
          </span>
        </button>
    </MgInput>)
  }
}
