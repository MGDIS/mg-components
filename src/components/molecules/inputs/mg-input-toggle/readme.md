# mg-input-toggle



<!-- Auto Generated Below -->


## Properties

| Property             | Attribute             | Description                                                                                                 | Type                        | Default                       |
| -------------------- | --------------------- | ----------------------------------------------------------------------------------------------------------- | --------------------------- | ----------------------------- |
| `disabled`           | `disabled`            | Define if input is disabled                                                                                 | `boolean`                   | `false`                       |
| `helpText`           | `help-text`           | Template to use for characters left sentence                                                                | `string`                    | `undefined`                   |
| `identifier`         | `identifier`          | Identifier is used for the element ID (id is a reserved prop in Stencil.js) If not set, it will be created. | `string`                    | `createID('mg-input-toggle')` |
| `invalid`            | `invalid`             | Force invalid component                                                                                     | `boolean`                   | `undefined`                   |
| `items` _(required)_ | --                    | Items are the possible options to select Required                                                           | `ToggleValue[] \| string[]` | `undefined`                   |
| `label` _(required)_ | `label`               | Input label Required                                                                                        | `string`                    | `undefined`                   |
| `labelHide`          | `label-hide`          | Define if label is visible                                                                                  | `boolean`                   | `false`                       |
| `labelOnTop`         | `label-on-top`        | Define if label is displayed on top                                                                         | `boolean`                   | `undefined`                   |
| `name`               | `name`                | Input name If not set the value equals the identifier                                                       | `string`                    | `this.identifier`             |
| `readonly`           | `readonly`            | Define if input is readonly                                                                                 | `boolean`                   | `false`                       |
| `required`           | `required`            | Define if input is required                                                                                 | `boolean`                   | `false`                       |
| `tooltip`            | `tooltip`             | Add a tooltip message next to the input                                                                     | `string`                    | `undefined`                   |
| `valid`              | `valid`               | Force valid component                                                                                       | `boolean`                   | `undefined`                   |
| `value`              | `value`               | Component value                                                                                             | `any`                       | `undefined`                   |
| `valuesSideBySide`   | `values-side-by-side` | Define if values are displaed side by side                                                                  | `boolean`                   | `false`                       |


## Events

| Event         | Description                     | Type               |
| ------------- | ------------------------------- | ------------------ |
| `valueChange` | Emmited event when value change | `CustomEvent<any>` |


## Dependencies

### Depends on

- [mg-tooltip](../../../atoms/mg-tooltip)
- [mg-icon](../../../atoms/mg-icon)
- [mg-input-title](../../../atoms/mg-input-title)
- [mg-character-left](../../../atoms/mg-character-left)

### Graph
```mermaid
graph TD;
  mg-input-toggle --> mg-tooltip
  mg-input-toggle --> mg-icon
  mg-input-toggle --> mg-input-title
  mg-input-toggle --> mg-character-left
  style mg-input-toggle fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
