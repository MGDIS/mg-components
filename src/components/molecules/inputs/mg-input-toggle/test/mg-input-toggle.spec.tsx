import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgInputToggle } from '../mg-input-toggle';
import { messages } from '../../../../../locales';
import { MgIcon } from '../../../../atoms/mg-icon/mg-icon';

const getPage = (args, slots?) => newSpecPage({
  components: [MgInputToggle, MgIcon],
  template: () => (<mg-input-toggle {...args}>{slots?.map(slot => slot())}</mg-input-toggle>)
});

describe('mg-input-toggle', () => {
  describe.each([
    undefined,
    [() => <span slot="item-1">Choix A</span>, () => <span slot="item-2">Choix B</span>],
    [() => <span slot="item-1"><mg-icon icon="cross"></mg-icon></span>, () => <span slot="item-2"><mg-icon icon="check"></mg-icon></span>]
  ])('template', slots => {
    test.each([
      {label: 'label', identifier: "identifier", items: [{title: 'batman', value: false}, {title: 'joker', value: true}]},
      {label: 'label', identifier: "identifier", items: [{title: 'batman', value: false}, {title: 'joker', value: true}], isIcon: true},
      {label: 'label', identifier: "identifier", items: [{title: 'batman', value: false}, {title: 'joker', value: true}], isOnOff: true},
      {label: 'label', identifier: "identifier", items: [{title: 'batman', value: false}, {title: 'joker', value: true}], labelOnTop: true},
      {label: 'label', identifier: "identifier", items: [{title: 'batman', value: false}, {title: 'joker', value: true}], labelHide: true},
      {label: 'label', identifier: "identifier", items: [{title: 'batman', value: false}, {title: 'joker', value: true}], required: true},
      {label: 'label', identifier: "identifier", items: [{title: 'batman', value: false}, {title: 'joker', value: true}], readonly: true},
      {label: 'label', identifier: "identifier", items: [{title: 'batman', value: false}, {title: 'joker', value: true}], disabled: true},
      {label: 'label', identifier: "identifier", items: [{title: 'batman', value: false}, {title: 'joker', value: true}], helpText: 'Hello joker'},
      {label: 'label', identifier: "identifier", items: [{title: 'batman', value: false}, {title: 'joker', value: true}], tooltip: "My Tooltip Message"},
      {label: 'label', identifier: "identifier", items: [{title: 'batman', value: false}, {title: 'joker', value: true}], tooltip: "My Tooltip Message", labelOnTop: true}
    ])('Should render with args %s:', async (args) => {
      const {root} = await getPage(args, slots);
      expect(root).toMatchSnapshot();
    });
  })

  test.each(["", undefined])('Should throw error with invalid label property : %s', async (value) => {
    try {
      await getPage({label:value, items: [{title: 'batman', value: false}, {title: 'joker', value: true}]});
    }
    catch (err) {
      expect(err.message).toMatch('prop "label" is required')
    }
  });

  test('Should throw an error with labelOnTop & labelHide set to true', async () => {
    try {
      await getPage({label: 'batman', labelOnTop: true, labelHide: true, items: ['batman', 'joker']});
    }
    catch (err) {
      expect(err.message).toMatch('<mg-input> prop "labelOnTop" must not be paired with the prop "labelHide"')
    }
  });

  test('Should throw an error with less than 2 items, case %s', async () => {
    try {
      await getPage({label: 'batman', items: [{title: 'batman', value: false}] });
    }
    catch (err) {
      expect(err.message).toMatch('<mg-input-toggle> prop "items" require 2 items.')
    }
  });

  test.each([
    [['batman', {title:'batman', value:'batman'}]],
    [['batman', {batman:'batman'}]],
    [[true, false]],
    [[1, 2]],
    [[true, 'batman']],
    [[{title:'batman', value:'batman'}, {batman:'batman'}]],
    [[{title:'batman', value: undefined}, {title:'batman', value: 'test'}]],
  ])('Should throw error with invalid items property : %s', async (items) => {
    try {
      await getPage({label:'Label', items});
    }
    catch (err) {
      expect(err.message).toMatch('<mg-input-toggle> prop "items" is required and all items must be the same type: ToggleValue.')
    }
  });


  test.each([
    {items: [{ title: 'batman', value: 'a'}, { title: 'joker', value: 'b' }], expected: 'b', value: undefined},
    {items: [{ title: 'batman', value: 'a'}, { title: 'joker', value: 'b' }], expected: 'a', value: 'b'},
    {items: [{ title: 'batman', value: 1}, { title: 'joker', value: 2 }], expected: 2, value: 1},
    {items: [{ title: 'batman', value: false },{ title: 'robin', value: true }], expected: false, value: true}
  ])('Should trigger events for items with inputValue : %s', async ({items, expected, value})=> {
    const args = {label: 'label', items , identifier: "identifier", helpText: "My help text", value};
    const page = await getPage(args);

    const button = page.doc.querySelector('button');

    jest.spyOn(page.rootInstance.valueChange, 'emit');

    button.dispatchEvent(new CustomEvent('focus', { bubbles: true }));
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot(); //Snapshot on focus

    button.dispatchEvent(new CustomEvent('click', { bubbles: true }));
    await page.waitForChanges();

    expect(page.rootInstance.valueChange.emit).toHaveBeenCalledWith(expected);
    expect(page.rootInstance.valueChange.emit).toHaveBeenCalledTimes(1);
  });

  test.each([
    {validity: true, valueMissing: false},
    {validity: false, valueMissing: true},
  ])('validity (%s), valueMissing (%s)', async ({validity, valueMissing})=> {
    const value = valueMissing ? undefined : true;
    const args = {label: 'label', items: [{title: 'batman', value: false}, {title: 'joker', value: true}], identifier: "identifier", patternErrorMessage: "Non", value};

    const page = await getPage(args);

    const button = page.doc.querySelector('button');

    button.dispatchEvent(new CustomEvent('blur', { bubbles: true }));
    await page.waitForChanges();

    if(validity) {
      expect(page.rootInstance.errorMessage).toBeUndefined();
    }
    else if (valueMissing){
      expect(page.rootInstance.errorMessage).toEqual(messages.errors.required);
    }
    expect(page.rootInstance.valid).toEqual(validity);
    expect(page.rootInstance.invalid).toEqual(!validity);
  });

  describe('keyboard', () => {
    test.each(['ArrowRight', 'ArrowLeft', 'ArrowUp', 'ArrowDown'])('should trigger toggle, case KeyDown %s', async (key) => {
      const args = {label: 'label', items: [{title: 'batman', value: false}, {title: 'joker', value: true}], identifier: "identifier"};

      const page = await getPage(args);

      const button = page.doc.querySelector('button');

      button.dispatchEvent(new CustomEvent('focus', { bubbles: true }));
      await page.waitForChanges();

      expect(page.rootInstance.value).toBeUndefined();

      button.dispatchEvent(new KeyboardEvent('keydown', {'code': key}));
      await page.waitForChanges();

      expect(page.rootInstance.value).toEqual("");
    })

    test.each(['Space', 'Enter'])('should NOT trigger toggle, case KeyDown %s', async (key) => {
      const args = {label: 'label', items: [{title: 'batman', value: false}, {title: 'joker', value: true}], identifier: "identifier"};

      const page = await getPage(args);

      const button = page.doc.querySelector('button');

      button.dispatchEvent(new CustomEvent('focus', { bubbles: true }));
      await page.waitForChanges();

      expect(page.rootInstance.value).toBeUndefined();

      button.dispatchEvent(new KeyboardEvent('keydown', {'code': key}));
      await page.waitForChanges();

      expect(page.rootInstance.value).toBeUndefined();
    })
  })
});
