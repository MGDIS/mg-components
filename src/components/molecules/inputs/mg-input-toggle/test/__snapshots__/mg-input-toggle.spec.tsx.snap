// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-input-toggle Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  disabled: true
}: 1`] = `
<mg-input-toggle>
  <!---->
  <div class="mg-input mg-input--toggle">
    <mg-input-title identifier="identifier">
      label
    </mg-input-title>
    <div class="mg-input__input-container">
      <div class="mg-input__input-container__input">
        <button class="mg-input-toggle__button mg-input-toggle__button--disabled" disabled="" id="identifier" role="switch" type="button">
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
        </button>
      </div>
    </div>
  </div>
</mg-input-toggle>
`;

exports[`mg-input-toggle Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  displayBothValues: true
}: 1`] = `
<mg-input-toggle>
  <!---->
  <div class="mg-input mg-input--toggle mg-input-toggle--display-both-values">
    <mg-input-title identifier="identifier">
      label
    </mg-input-title>
    <div class="mg-input__input-container">
      <div class="mg-input__input-container__input">
        <button class="mg-input-toggle__button" id="identifier" role="switch" type="button">
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
        </button>
      </div>
    </div>
  </div>
</mg-input-toggle>
`;

exports[`mg-input-toggle Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  helpText: 'Hello joker'
}: 1`] = `
<mg-input-toggle>
  <!---->
  <div class="mg-input mg-input--toggle">
    <mg-input-title identifier="identifier">
      label
    </mg-input-title>
    <div class="mg-input__input-container">
      <div class="mg-input__input-container__input">
        <button class="mg-input-toggle__button" id="identifier" role="switch" type="button">
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
        </button>
      </div>
      <div class="mg-input__input-container__help-text" id="identifier-help-text">
        Hello joker
      </div>
    </div>
  </div>
</mg-input-toggle>
`;

exports[`mg-input-toggle Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  labelHide: true
}: 1`] = `
<mg-input-toggle>
  <!---->
  <div class="mg-input mg-input--toggle">
    <mg-input-title class="sr-only" identifier="identifier">
      label
    </mg-input-title>
    <div class="mg-input__input-container">
      <div class="mg-input__input-container__input">
        <button class="mg-input-toggle__button" id="identifier" role="switch" type="button">
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
        </button>
      </div>
    </div>
  </div>
</mg-input-toggle>
`;

exports[`mg-input-toggle Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  labelOnTop: true
}: 1`] = `
<mg-input-toggle>
  <!---->
  <div class="mg-input mg-input--label-on-top mg-input--toggle">
    <mg-input-title identifier="identifier">
      label
    </mg-input-title>
    <div class="mg-input__input-container">
      <div class="mg-input__input-container__input">
        <button class="mg-input-toggle__button" id="identifier" role="switch" type="button">
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
        </button>
      </div>
    </div>
  </div>
</mg-input-toggle>
`;

exports[`mg-input-toggle Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  readonly: true
}: 1`] = `
<mg-input-toggle>
  <!---->
  <div class="mg-input mg-input--readonly mg-input--toggle">
    <mg-input-title identifier="identifier">
      label
    </mg-input-title>
    <div class="mg-input__input-container">
      <strong></strong>
    </div>
  </div>
</mg-input-toggle>
`;

exports[`mg-input-toggle Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  required: true
}: 1`] = `
<mg-input-toggle>
  <!---->
  <div class="mg-input mg-input--toggle">
    <mg-input-title identifier="identifier" required="">
      label
    </mg-input-title>
    <div class="mg-input__input-container">
      <div class="mg-input__input-container__input">
        <button class="mg-input-toggle__button" id="identifier" role="switch" type="button">
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
        </button>
      </div>
    </div>
  </div>
</mg-input-toggle>
`;

exports[`mg-input-toggle Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  tooltip: 'My Tooltip Message'
}: 1`] = `
<mg-input-toggle>
  <!---->
  <div class="mg-input mg-input--toggle">
    <mg-input-title identifier="identifier">
      label
    </mg-input-title>
    <div class="mg-input__input-container">
      <div class="mg-input__input-container__input">
        <button class="mg-input-toggle__button" id="identifier" role="switch" type="button">
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
        </button>
        <mg-tooltip identifier="identifier-tooltip" message="My Tooltip Message">
          <mg-icon icon="info"></mg-icon>
        </mg-tooltip>
      </div>
    </div>
  </div>
</mg-input-toggle>
`;

exports[`mg-input-toggle Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  tooltip: 'My Tooltip Message',
  labelOnTop: true
}: 1`] = `
<mg-input-toggle>
  <!---->
  <div class="mg-input mg-input--label-on-top mg-input--toggle">
    <mg-input-title identifier="identifier">
      label
    </mg-input-title>
    <mg-tooltip identifier="identifier-tooltip" message="My Tooltip Message">
      <mg-icon icon="info"></mg-icon>
    </mg-tooltip>
    <div class="mg-input__input-container">
      <div class="mg-input__input-container__input">
        <button class="mg-input-toggle__button" id="identifier" role="switch" type="button">
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
        </button>
      </div>
    </div>
  </div>
</mg-input-toggle>
`;

exports[`mg-input-toggle Should render with args { label: 'label', identifier: 'identifier', items: [Array] }: 1`] = `
<mg-input-toggle>
  <!---->
  <div class="mg-input mg-input--toggle">
    <mg-input-title identifier="identifier">
      label
    </mg-input-title>
    <div class="mg-input__input-container">
      <div class="mg-input__input-container__input">
        <button class="mg-input-toggle__button" id="identifier" role="switch" type="button">
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
        </button>
      </div>
    </div>
  </div>
</mg-input-toggle>
`;

exports[`mg-input-toggle Should trigger events for items with inputValue : { items: [Array], expected: 'a', value: 'b' } 1`] = `
<mg-input-toggle value="b">
  <!---->
  <div class="mg-input mg-input--toggle mg-input--toggle-active">
    <mg-input-title identifier="identifier">
      label
    </mg-input-title>
    <div class="mg-input__input-container">
      <div class="mg-input__input-container__input">
        <button aria-checked="b" class="mg-input-toggle__button" id="identifier" role="switch" type="button">
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
        </button>
      </div>
      <div class="mg-input__input-container__help-text" id="identifier-help-text">
        My help text
      </div>
    </div>
  </div>
</mg-input-toggle>
`;

exports[`mg-input-toggle Should trigger events for items with inputValue : { items: [Array], expected: 2, value: 1 } 1`] = `
<mg-input-toggle value="1">
  <!---->
  <div class="mg-input mg-input--toggle">
    <mg-input-title identifier="identifier">
      label
    </mg-input-title>
    <div class="mg-input__input-container">
      <div class="mg-input__input-container__input">
        <button aria-checked="1" class="mg-input-toggle__button" id="identifier" role="switch" type="button">
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
        </button>
      </div>
      <div class="mg-input__input-container__help-text" id="identifier-help-text">
        My help text
      </div>
    </div>
  </div>
</mg-input-toggle>
`;

exports[`mg-input-toggle Should trigger events for items with inputValue : { items: [Array], expected: false, value: true } 1`] = `
<mg-input-toggle value="">
  <!---->
  <div class="mg-input mg-input--toggle mg-input--toggle-active">
    <mg-input-title identifier="identifier">
      label
    </mg-input-title>
    <div class="mg-input__input-container">
      <div class="mg-input__input-container__input">
        <button aria-checked="" class="mg-input-toggle__button" id="identifier" role="switch" type="button">
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
          <span aria-hidden="true" class="mg-input-toggle__item-container"></span>
        </button>
      </div>
      <div class="mg-input__input-container__help-text" id="identifier-help-text">
        My help text
      </div>
    </div>
  </div>
</mg-input-toggle>
`;
